### PR TITLE
feat: authentication - user login

### PR DESCRIPTION
**Pull request by ZEEZKING**
### Description
Implement a robust login feature to streamline user access and authentication processes. This feature should encompass secure user logins, password protection, and account security measures.

### Acceptance Criteria
 

- [ ] The endpoint must return a 422 error for missing or invalid input with an appropriate error message.
- [ ]  A 401 status code should be returned for invalid credentials with an appropriate error message.
- [ ]  A successful login should return a 200 response with success message
- [ ]  User's credentials transmitted securely to the Backend
- [ ]  Once validation checks are passed, JWT token is generated for the user encoded using the user mail, user id, and first and last names and returned as part of the response payload

### Expected Outcome

- The user should be redirected to the subsequent page if successful, or receive an error message if unsuccessful
- Users are informed of login activity.

### Requirements

- Corresponding database fields for:
- Email Address
- Password

### User Login [POST] /api/v1/auth/login

1. Login Endpoint

-  Given a request with valid user details (ieemail, password), when the user logs in, then the system should return the user data, access and refresh token with a 200 success status code.

 2. Incorrect credentials

- Given the input does not match with the credentials on the database, then the system should return a 400 bad request error status with an appropriate error message
**Request**

```
POST /api/v1/auth/login
{
  "email": String,
  "password": String
}
```
**Successful Response**

```
{
  "message": String,
  "data":
       "user": {
            "id": String,
            "email": String,
            "role": String,
            "signup_type": String,
            "is_active": Boolean,
            "is_verified": Boolean,
            "created_at": Date,
            "updated_at": Date,
            "last_login-at": Date
          },
      access_token: String,
      refresh_token: String
}
```
**Error Response**

```
{
  "message": String,
  "error": String,
  "status_code": Int,
}
```

### CODE REVIEW

**Comment by buka4rill** [comment]
CleanShot 2024-07-24 at 10 33 19
Wrong branch naming

**Comment by buka4rill** [comment]
rename the PR to 
`feat: authentication - user login`

**Comment by buka4rill** [comment]
Your team leads have to review first

File: src/Hng.Application.Test/Features/UserManagement/LoginUserQueryShould.cs [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+                Email = "test@example.com",
+                Password = "password"
+            });
+
+            // Act
+            var result = await handler.Handle(command, CancellationToken.None);
+
+            // Assert
+            Assert.NotNull(result);
+            //Assert.Equal("Login successful", result.Message);
```
Comment (NdubuisiJr) [inline src/Hng.Application.Test/Features/UserManagement/LoginUserQueryShould.cs:71]: Remove commented out code

File: src/Hng.Application.Test/Features/UserManagement/LoginUserQueryShould.cs [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+            Assert.NotNull(result.Data);
+            Assert.Equal(user.Email, result.Data.Email);
+        }
+
+
+        [Fact]
+        public async Task ReturnNullForInvalidCredentials()
+        {
+            // Arrange
+            var user = new User
```
Comment (NdubuisiJr) [inline src/Hng.Application.Test/Features/UserManagement/LoginUserQueryShould.cs:82]: Since this object is reused, you can initialize it in the constructor

**Review by NdubuisiJr — CHANGES_REQUESTED** [review]
Almost done. Resolve the comments below.

File: src/Hng.Application/Features/UserManagement/Dtos/UserLoginResponseDto.cs [hunk 3]
```diff
@@ -0,0 +1,14 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Hng.Application.Features.UserManagement.Dtos
+{
+    public class UserLoginResponseDto
+    {
```
Comment (NdubuisiJr) [inline src/Hng.Application/Features/UserManagement/Dtos/UserLoginResponseDto.cs:10]: This should the `message` property.

**Review by Godhanded — CHANGES_REQUESTED** [review]

File: src/Hng.Web/Controllers/UserController.cs [hunk 4]
```diff
@@ -47,5 +47,26 @@ public async Task<ActionResult<UserDto>> CreateUser([FromBody] UserCreationDto b
...
+            var command = new CreateUserLoginCommand(loginRequest);
+            var response = await _mediator.Send(command);
+
+            if (response.Data == null)
+            {
+                return BadRequest(new
+                {
+                    message = "Invalid credentials",
+                    error = "Invalid email or password.",
+                    status_code = StatusCodes.Status400BadRequest
```
Comment (NdubuisiJr) [inline src/Hng.Web/Controllers/UserController.cs:65]: The response here should not be `Bad request` it should be `401` Unauthorized

**Review by NdubuisiJr — CHANGES_REQUESTED** [review]

File: src/Hng.Web/Controllers/UserController.cs (hunk 4 above)
Comment (NdubuisiJr) [reply src/Hng.Web/Controllers/UserController.cs:65]: Also this should not be in user controller. You route will now be `/api/v1/users/login`. Is this how it's supposed to be?

Comment (ZEEZKING) [reply src/Hng.Web/Controllers/UserController.cs:65]: ok sir buh i dont understand very well sir

**Pull request closed without merging**